### PR TITLE
ENH: interpolate: Make an improvement to support complete coefficients (the second way of #14773)

### PR DESCRIPTION
**Pull request by zhaog6**
#### Reference issue

To solve #14681 
The first implementation way: #14773 

#### What does this implement/fix?

Make some improvements with the function `lagrange` ( the second implementation way of #14773)

#### Additional information

Usage:
```python
from scipy.interpolate import lagrange

x = np.arange(3)
y = np.ones(3)
y *= 0.5
pl = lagrange(x, y, reduce=False)
print(f"Zero-filled coefficients: {pl.coord}")
print(f"Standard coefficients: {pl.coeffs}")

pl = lagrange(x, y, reduce=True)
print(f"Non-zero-filled coefficients: {pl.coord}")
print(f"Standard coefficients: {pl.coeffs}")
```

### CODE REVIEW

File: scipy/interpolate/interpolate.py [hunk 1]
```diff
@@ -22,12 +22,13 @@
 from ._bsplines import make_interp_spline, BSpline
 
 
-def lagrange(x, w):
+def lagrange(x, y, reduce=True):
```
Comment (ev-br) [inline scipy/interpolate/interpolate.py:25]: Can't just rename arguments like this : it's breaking back compat for anybody who uses a keywords arg, `lagrage(x, w=...)`.

File: scipy/interpolate/interpolate.py [hunk 2]
```diff
@@ -22,12 +22,13 @@
...
     r"""
-    Return a Lagrange interpolating polynomial.
+    poly = lagrange(x, y)
```
Comment (ev-br) [inline scipy/interpolate/interpolate.py:27]: This line is breaking the docstring rendering: the first line is a full sentence of a short description of a function. Best roll back this change.

File: scipy/interpolate/interpolate.py [hunk 3]
```diff
@@ -22,12 +22,13 @@
...
+    Return a Lagrange interpolating polynomial
 
-    Given two 1-D arrays `x` and `w,` returns the Lagrange interpolating
-    polynomial through the points ``(x, w)``.
+    Given two 1-D arrays `x` and `y,` returns the Lagrange interpolating
+    polynomial through the points ``(x, y)``.
```
Comment (ev-br) [inline scipy/interpolate/interpolate.py:31]: `w` should be back here, too.

File: scipy/interpolate/interpolate.py [hunk 4]
```diff
@@ -36,13 +37,21 @@ def lagrange(x, w):
     ----------
     x : array_like
         `x` represents the x-coordinates of a set of datapoints.
-    w : array_like
-        `w` represents the y-coordinates of a set of datapoints, i.e., f(`x`).
+    y : array_like
+        `y` represents the y-coordinates of a set of datapoints, i.e., f(`x`).
+    reduce : bool
```
Comment (ev-br) [inline scipy/interpolate/interpolate.py:42]: `bool, optional`

File: scipy/interpolate/interpolate.py [hunk 5]
```diff
@@ -36,13 +37,21 @@ def lagrange(x, w):
...
+        `reduce = True` represents use reduced Lagrange coefficients.
+        `reduce = False` represents use complete Lagrange coefficients (the size
+        of the coefficient array equals the size of x).
+        Default is `True`.
```
Comment (ev-br) [inline scipy/interpolate/interpolate.py:46]: TBH, I've no idea what reduced or complete Lagrange coefficients are. This does not seem to be a term my google-fu finds either. Best define what the switch does like you did in brackets and definitely leave the default back-compatible.

File: scipy/interpolate/interpolate.py [hunk 6]
```diff
@@ -36,13 +37,21 @@ def lagrange(x, w):
...
 
     Returns
     -------
     lagrange : `numpy.poly1d` instance
         The Lagrange interpolating polynomial.
+    coeffs : ndarray
+        If `reduce = True`, only `lagrange` is returned.
+        Otherwise, a tuple `(lagrange, coeffs)` is returned.
```
Comment (ev-br) [inline scipy/interpolate/interpolate.py:54]: Yuck, variable-length returns. Makes me question the whole premise for the patch TBH.

File: scipy/interpolate/interpolate.py [hunk 7]
```diff
@@ -75,19 +84,26 @@ def lagrange(x, w):
...
-        pt = poly1d(w[j])
+        pt = poly1d(y[j])
         for k in range(M):
             if k == j:
                 continue
             fac = x[j]-x[k]
             pt *= poly1d([1.0, -x[k]])/fac
         p += pt
+    diff = M - len(p.coeffs)
+    if reduce is False:
```
Comment (ev-br) [inline scipy/interpolate/interpolate.py:100]: `if not reduce`

File: scipy/interpolate/interpolate.py [hunk 8]
```diff
@@ -75,19 +84,26 @@ def lagrange(x, w):
...
     for j in range(M):
-        pt = poly1d(w[j])
+        pt = poly1d(y[j])
         for k in range(M):
             if k == j:
                 continue
             fac = x[j]-x[k]
             pt *= poly1d([1.0, -x[k]])/fac
         p += pt
+    diff = M - len(p.coeffs)
```
Comment (ev-br) [inline scipy/interpolate/interpolate.py:99]: This is only needed in the `if not reduce` branch, so best move this line over there

File: scipy/interpolate/interpolate.py [hunk 9]
```diff
@@ -75,19 +84,26 @@ def lagrange(x, w):
...
+    if reduce is False:
+        if diff > 0:
+            tmp = np.zeros(diff)
+            coef = np.append(tmp, p.coeffs)
+            return p, coef
```
Comment (ev-br) [inline scipy/interpolate/interpolate.py:104]: This all needs a comment. The logic is not hard to parse when you look at it, but the first glance is puzzling, so it needs a short comment.

File: scipy/interpolate/tests/test_interpolate.py [hunk 10]
```diff
@@ -753,13 +753,21 @@ def test_read_only(self):
...
-        p = poly1d([5,2,1,4,3])
+        p = poly1d([5, 2, 1, 4, 3])
         xs = np.arange(len(p.coeffs))
         ys = p(xs)
-        pl = lagrange(xs,ys)
-        assert_array_almost_equal(p.coeffs,pl.coeffs)
+        pl = lagrange(xs, ys)
+        assert_array_almost_equal(p.coeffs, pl.coeffs)
+
+    def test_reduced_coeffs(self):
```
Comment (ev-br) [inline scipy/interpolate/tests/test_interpolate.py:763]: The test also needs a comment of what exactly it tests.

File: scipy/interpolate/tests/test_interpolate.py [hunk 11]
```diff
@@ -753,13 +753,21 @@ def test_read_only(self):
...
+        a = np.array([0, 2, 1])
+        p = poly1d(a)
+        x = np.arange(len(a))
+        y = p(x)
+        pl = lagrange(x, y, reduce=False)
```
Comment (ev-br) [inline scipy/interpolate/tests/test_interpolate.py:768]: Since you're adding a boolean flag, both values need to be tested.

File: scipy/interpolate/tests/test_interpolate.py [hunk 12]
```diff
@@ -753,13 +753,21 @@ def test_read_only(self):
...
+        assert_array_almost_equal(a, pl[1])
+        assert_array_almost_equal(p.coeffs, pl[0].coeffs)
```
Comment (ev-br) [inline scipy/interpolate/tests/test_interpolate.py:770]: Nit: in new code, prefer `assert_allclose` with explicit tolerance (`atol=1e-14` or some such for testing fp equality)

**Review by ev-br — CHANGES_REQUESTED** [review]
I've left a review with several technical things to address if this goes in. The patch itself is a very good starting point --- However, the more I read this, the more I get to question if this is more hassle than it's worth. 

First of all, where does the initial request come from? Granted,  it seems to be a valid feature request for a somewhat one-off feature. One can argue that the feature should live in a user code or in the library code, and it's basically a wash. So it boils down to the question of the maintenance burden. 
This patch shows that it can be done without too much difficulty, and the maintenance burden
increase in negligible, so it can probably go in. The boolean flag controlled returns are ugly and are best avoided, but here I understand it's either that, or reject the whole thing. 

All in all, since you've already done the work and if you're willing to finish it up and make sure the backcompat is preserved, this may go in.

**Comment by zhaog6** [comment]
> Granted, it seems to be a valid feature request for a somewhat one-off feature. One can argue that the feature should live in a user code or in the library code

Yes, Users only need to write a few lines of code if the feature live in a user code (I‘ve also been thinking about it). I think that if the patch is easy to maintain, can be more user-friendly and has a clear research and development purposes, it is worth adding the functionality.

**Comment by zhaog6** [comment]
> All in all, since you've already done the work and if you're willing to finish it up and make sure the backcompat is preserved, this may go in.

Thanks your review and all advices, I'll make the changes later.

File: scipy/interpolate/interpolate.py (hunk 6 above)
Comment (zhaog6) [reply scipy/interpolate/interpolate.py:54]: > Yuck, variable-length returns. Makes me question the whole premise for the patch TBH.

It seems difficult to ensure the consistent variable length to return (if inheritance from `poly1d` is not used) because the return value `p` in the original code is an object of  the class `poly1d` and for back-compatibility reasons.

File: scipy/interpolate/tests/test_interpolate.py [hunk 13]
```diff
@@ -753,13 +753,21 @@ def test_read_only(self):
...
-
     def test_lagrange(self):
-        p = poly1d([5,2,1,4,3])
+        p = poly1d([5, 2, 1, 4, 3])
         xs = np.arange(len(p.coeffs))
         ys = p(xs)
-        pl = lagrange(xs,ys)
-        assert_array_almost_equal(p.coeffs,pl.coeffs)
+        pl = lagrange(xs, ys)
+        assert_array_almost_equal(p.coeffs, pl.coeffs)
```
Comment (zhaog6) [inline scipy/interpolate/tests/test_interpolate.py:761]: > Since you're adding a boolean flag, both values need to be tested.

Because there is a test case with the default parameter in the `def test_lagrange(self)`. Do I still need to add the `pl = lagrange(xs, ys, reduce=True)` explicitly?

Comment (ev-br) [reply scipy/interpolate/tests/test_interpolate.py:761]: Ok, then please add a comment to this effect

File: scipy/interpolate/interpolate.py (hunk 6 above)
Comment (zhaog6) [reply scipy/interpolate/interpolate.py:54]: @ev-br I try to ensure the consistent length of the return value, but I feel it thorny. Do you have any good way or suggestion?

Comment (ev-br) [reply scipy/interpolate/interpolate.py:54]: Nope, as long as it returns a poly1d object, which is frozen by backcompat

**Comment by zhaog6** [comment]
Dear @ev-br 
Please ignore all commits above because these commits are too messy. The following is a brand new submission. Here the parameter `reduce` is not added and keep the original parameter options, which simplifies the code implementation. Looking forward to your review again. Thanks.
